Kafka codec

adamkotwasinski opened the pull request:
*Description*: Placeholder for Kafka codec, API inspired on some level by Mongo codec (internals partially by Redis though)
*Risk Level*: medium
*Testing*: so far manual, by setting up Kafka cluster & Kafka clients (I keep filter code that's relatively trivial and not part of this PR)

Recommended (?) order of reading:
* `codec.h` - abstract codecs (there will be one for kafka request & response - to be used in `onData`/`onWrite` filter methods)
* `request_codec.h / .cc` - kafka request encoder&decoder for requests - decoder notifies listeners (similar to mongo) and uses a parser-loop to handle complex messages
* `parser.h` - parser is a parser (the name isn't perfect) - it's a stateful thingy that keeps current buffering state, and can tell if it's done (returns a message) or should be changed (nextParser)
* (now jump a bit to the bottom) `kafka_types.h` - kafka data types
* `kafka_protocol.h` - constants coming from kafka (request-only now)
* `serialization.h` - how `char*` gets translated into kafka types + complex buffer template (that combines N buffers, and then combines their result into one return object) and encoder
* `kafka_request.h / .cc`:
** request header (4 fields present in *every* msg)
** request context keeping that header that's passed in the parser chain mentioned above
** request mapper (basically a mapper from request_type x request_version -> parser)
** message start parser (consumes message length)
** header parser (consumes request header)
** some request parsers (keys 0-9 with versions up to Kafka 0.11)
** (at the bottom) `UnknownRequest` for requests that couldn't be recognised by mapper

Tests:
* `serialization_test`- trivial kafka types, checked in both ways: feed whole buffer at once & trickle by one byte (to ensure state is being kept by buffer properly - compare with redis filter that appears to be doing the same)
* `kafka_request_test` - mapper testing; the details of parse loop `start parser` -> `header parser` -> `specific parser`; 
* `request_codec_test` - full tests for each of currently implemented request type/version combos

rshriram commented on source/extensions/filters/network/kafka/kafka_request.h:
```diff
@@ -0,0 +1,1316 @@
+
+// === REQUEST HEADER ==========================================================
+
+struct RequestHeader {
```
is it really called request in Kafka? isn't it more of a message/pub/sub or operation?

adamkotwasinski replied:
the underlying entities are really called requests/responses
see https://kafka.apache.org/protocol#protocol_common & https://kafka.apache.org/protocol#protocol_messages (and corresponding classes in java client like `ProduceRequest`)

mattklein123 commented:
Sorry for the delay. I'm still planning on reviewing this. I'm behind.

adamkotwasinski commented on source/extensions/filters/network/kafka/serialization.h:
```diff
@@ -0,0 +1,772 @@
+ * Composes several buffers into one.
+ * The returned value is constructed via { buffer1.get(), buffer2.get() ... }
+ */
+template <typename RT, typename...> class CompositeBuffer;
```
I just learned I can specialize this like
```
template <class... Ts> struct tuple {};

template <class T, class... Ts>
struct tuple<T, Ts...> : tuple<Ts...> {
  tuple(T t, Ts... ts) : tuple<Ts...>(ts...), tail(t) {}
  T tail;
};
```

adamkotwasinski replied:
The most-offending `Composite` deserializer template has been removed and replaced with explicit domain deserializers.
All other deserializers are subclasses of `Deserializer<T>` interface and deserialize into instance of T.

The encoding template functions have been documented - in general I wanted to have single method that could work as `encode(string)`, `encode(int8)`, `encode(bool)` vs having `encodeString`/`encodeInt8`/`encodeBool` as IMHO it makes it harder to make generic code

mattklein123 replied:
Sounds good. Let's see how it looks on the next pass.

mattklein123 commented:
@adamkotwasinski sorry for the really long delay. I have time to review this now. This is too much code to review in a single PR. Can you potentially cut this down to whatever infrastructure is required for a few Kafka messages and an incomplete codec and associated tests? Then we can review just that to make sure we are on the same page and go from there?

adamkotwasinski commented:
@mattklein123 Done! Left only one request type, but it's complex enough to see how the versioning / changing structure is handled.

mattklein123 commented on source/extensions/filters/network/kafka/kafka_request.h:
```diff
@@ -0,0 +1,398 @@
+  RequestHeader request_header_;
+};
+
+// === OFFSET COMMIT (8) =======================================================
```
In general, I think this code would benefit from starting to split it into multiple files. Can we potentially have a file for base objects/classes, and then a file per message type since they seem to be fairly large? I think that would make the code easier to read and review as it grows.

adamkotwasinski replied:
naming them `request_type.h` and putting them in `messages` subdirectory
I envision one header will contain both request & response for the same type

adamkotwasinski commented on source/extensions/filters/network/kafka/kafka_types.h:
```diff
@@ -0,0 +1,31 @@
+namespace NetworkFilters {
+namespace Kafka {
+
+typedef int8_t INT8;
```
Sure!
My rationale was that I wanted to use "kafka types" such as `INT8` when I use things coming from Kafka itself. e.g. request's api_version is INT16, but I could have made an explicit `typedef uint16_t api_version` too

adamkotwasinski commented on source/extensions/filters/network/kafka/request_codec.cc:
```diff
@@ -0,0 +1,62 @@
+  Buffer::OwnedImpl data_buffer;
+  INT32 data_len = encoder.encode(message, data_buffer); // encode data computing data length
+  encoder.encode(data_len, output_);                     // encode data length into result
+  output_.add(data_buffer);                              // copy data into result
```
I would like to defer this problem until we decide how actually we are going to serialize/deserialize the data. First let's solve how are we going to manage `encode` / `decode` operations, `computeSize` will be yet another that will follow precisely the same pattern (after all our requests/responses are trivial trees - so let's decide on tooling to traverse the trees first).
Makes sense?

mattklein123 replied:
Yup that's fine, I would just put a TODO comment to remind us of a potential optimization later.

adamkotwasinski commented on source/extensions/filters/network/kafka/kafka_protocol.h:
```diff
@@ -0,0 +1,123 @@
+namespace Kafka {
+
+// from http://kafka.apache.org/protocol.html#protocol_api_keys
+enum RequestType : INT16 {
```
but then I cannot use `RequestType::OffsetFetch` as INT16 in my OffsetFetchRequest constructors .... or am I missing something?

mattklein123 replied:
Can't you just use `int16_t`? All I'm saying is to use the system types which are really clear? I don't see the point in the typedef when we use these types all over the code base?

adamkotwasinski commented on source/extensions/filters/network/kafka/kafka_request.h:
```diff
@@ -0,0 +1,398 @@
+
+// === VECTOR ==================================================================
+
+template <typename T> std::ostream& operator<<(std::ostream& os, const std::vector<T>& arg) {
```
moved to `debug_helpers.h`

adamkotwasinski commented on source/extensions/filters/network/kafka/messages/offset_commit.h:
```diff
@@ -0,0 +1,292 @@
+    return consumed;
+  }
+  bool ready() const { return partitions_.ready(); }
+  OffsetCommitTopic get() const { return {topic_.get(), partitions_.get()}; }
```
I think it starts to be pretty obvious here that `OffsetCommitPartitionV0Buffer`, `OffsetCommitPartitionV1Buffer`, `OffsetCommitTopicV0Buffer`, `OffsetCommitTopicV1Buffer` & co. are very similar to each other, being composed of 2 or 3 or 4 sub-deserializers, that then take the deserialization result, and just bundle them up into resulting `T` from `Deserializer<T>` via `{ .... }` constructor

I think we can attack it in a few different ways:
a) keep doing the same stuff, and just cope with the code bloat
b) introduce a new class `CompositeDeserializer` that would take a vector (tuple?) of sub-deserializers, something like

```
class CompositeDeserializer : public Deserializer<T> {
private:
  vector<Deserializer> delegates_; // I would be storing Deserializer<int8> next to Deserializer<string> here, so I might need a pointer I guess
public:
  bool ready() { return delegates_.end()->ready() } // same thing, ready when last deserializer is full
  size_t feed(const char*& buffer, uint64_t& remaining) {
    for (Deserializer &d : delegates) {
      d.feed(data, remaining)
    }
  }

  T get() {
    vector<void*> results = // get result from each of delegates and put them in vector
    return T{results} // needs a constructor that would downcast each of elements from void* to proper type
   }
}
```
What I don't like here I have `void*` flowing around requiring downcasting.

*c)* 
Common super classes looking like "deserializer composed of 2 deserializers"
```
class CompositeDeserializerComposedOf2Deserializers<T, Delegate1T, Delegate2T> {
private:
  Delegate1T d1;
  Delegate2T d2;
public:
  bool ready() { return d2.ready() }
  size_t feed (....) {
     d1.feed(....)
     d2.feed(....)
  }
  T get() {
    return { d1.get(), d2.get() };
  }
}
```
drawbacks : it's the template solution back again ;)

d) bonus consideration: make `OffsetCommitRequest` abstract and add `OffsetCommitRequestV0`, `OffsetCommitRequestV1` concrete classes - it could alleviate some of problems present in b) - as the constructor would always know that field 0 is e.g. topic name; instead of needing to check with `api_version` first

mattklein123 replied:
Hmm yeah, I think I'm starting to see the issue here, agree that b) is not great. I'm not completely opposed to templates, if you think it's the best way. I would pick what you think is best for now (with an eye towards trying to avoid complex templates if possible) and just make sure all of the header files are really well commented. Then I can take another pass? Sound good?

adamkotwasinski replied:
Sure! Thank you - I think it's visible the most when we actually have `N` semi-similar classes, then the benefit of (templated) superclass becomes appearent.

adamkotwasinski commented on source/extensions/filters/network/kafka/messages/offset_commit.h:
```diff
@@ -131,156 +131,48 @@ class OffsetCommitRequest : public Request {
+// Deserializes bytes into OffsetCommitRequest (api version 0): group_id, topics (v0)
+class OffsetCommitRequestV0Deserializer
+    : public CompositeDeserializerWith2Delegates<OffsetCommitRequest, StringDeserializer, OffsetCommitTopicV0ArrayBuffer> {};
 
```
one can write:
```
class OffsetCommitRequestV0Deserializer:
public CompositeDeserializerWith2Delegates<
  OffsetCommitRequest,
  StringDeserializer,
  ArrayDeserializer<
    OffsetCommitTopic,
    CompositeDeserializerWith2Delegates<
      OffsetCommitTopic,
      StringDeserializer,
      ArrayDeserializer<
        OffsetCommitPartition,
        CompositeDeserializerWith3Delegates<
          OffsetCommitPartition,
          Int32Deserializer,
          Int64Deserializer,
          NullableStringDeserializer
>>>>> {}
```
but I think it makes it even worse to understand

mattklein123 replied:
Yes that is not good. At this point, please just make the code is readable as you think it can be, while trying to remove egregious templates. When this is done I will take another pass and offer different suggestions if I see them. It will be much easier to do this with all the other changes I have asked for such as comments, etc.

adamkotwasinski replied:
The remaining big-template classes will be present only in `messages/*.h` files, as we need to define the structure somehow.
But in the end they are not very 

The end code is actually pretty similar to how Java client code looks like ( https://github.com/apache/kafka/blob/2.1/clients/src/main/java/org/apache/kafka/common/requests/OffsetCommitRequest.java#L58 ) but Java without templates forces them to use raw Objects flying around and (nicely hidden) downcasts just like here https://github.com/apache/kafka/blob/2.1/clients/src/main/java/org/apache/kafka/common/protocol/types/Type.java#L211 

Other changes like comments, splits, renames etc. have been already applied.

mattklein123 replied:
@adamkotwasinski OK so the PR is ready for another pass?

adamkotwasinski replied:
Yes

mattklein123 replied:
Can you fix DCO before I start reviewing again?

adamkotwasinski replied:
done

mattklein123 commented on source/extensions/filters/network/kafka/kafka_request.h:
```diff
@@ -0,0 +1,337 @@
+   * Consumes INT32 bytes as request length and updates the context with that value
+   * @return RequestHeaderParser instance to process request header
+   */
+  ParseResponse parse(const char*& buffer, uint64_t& remaining);
```
For overridden functions:

1) Use the `override` keyword
2) Don't duplicate the doc comment from the interface header
3) Precede the overrides of the comments with a comment similar to `// Extensions::Networkfilter::Kafka::Parser`

Please audit for this elsewhere.

mattklein123 commented on source/extensions/filters/network/kafka/kafka_request.h:
```diff
@@ -0,0 +1,337 @@
+ * @param BT deserializer type corresponding to request class (should be subclass of
+ * Deserializer<RT>)
+ */
+template <typename RT, typename BT> class RequestParser : public Parser {
```
nit: Spell out RT, BT, etc. same elsewhere.

mattklein123 commented on source/extensions/filters/network/kafka/request_codec.h:
```diff
@@ -0,0 +1,89 @@
+public:
+  virtual ~RequestCallback() = default;
+
+  virtual void onMessage(MessageSharedPtr) PURE;
```
doc comment

adamkotwasinski commented on source/extensions/filters/network/kafka/kafka_types.h:
```diff
@@ -0,0 +1,45 @@
+namespace Kafka {
+
+/**
+ * Represents a sequence of characters or null. For non-null strings, first the length N is given as
```
good point!, I can remove that in this header, this stuff is going to be present in `deserialization.h` anyways (as there I want to state why I'm parsing the way I am)

mattklein123 commented on source/extensions/filters/network/kafka/serialization_composite.h:
```diff
@@ -0,0 +1,133 @@
+namespace NetworkFilters {
+namespace Kafka {
+
+/**
```
Why are these templates actually needed? AFAICT all they do is string together multiple `feed` calls. Can't `feed` be an interface method, and then there can be a class that takes a list/vector of feeders and generically does what all of this template code does?

adamkotwasinski replied:
I think what you described is basically point `b)` of https://github.com/envoyproxy/envoy/pull/4950#discussion_r234224539

The `feed` part is not a problem, but I'm getting problems with constructing results in `get()`.
If I keep a vector of `Deserializer`-s then their return type is going to have to be something like `void*` or anything sufficiently generic, and then I'd need to convert these results into `Request`'s constructor arguments.

Basically would want to have something like
```
std::vector<Deserializer<?>> delegates_;

ReturnType get() const {
  return { delegates_[0].get(), delegates_[1].get(), delegates_[2].get() .... };
}
```

As a minimum I will `feed` will be changed to be more generic.

When it comes to `get`, I will take a look into possibility of having an array instead of vector - this way I think I'll be able to do some templating to change `array<N>{ deserializer1, ....}` into argument list.

mattklein123 replied:
OK I see the problem. Alright do what you can and I will take another pass in the next round. Sorry for not seeing the get() issue.

mattklein123 commented on source/extensions/filters/network/well_known_names.h:
```diff
@@ -30,6 +30,8 @@ class NetworkFilterNameValues {
   // Authorization filter
   const std::string ExtAuthorization = "envoy.ext_authz";
+  // Kafka filter
+  const std::string Kafka = "envoy.filters.network.kafka";
```
remove from this PR

mattklein123 commented:
@adamkotwasinski one other random high level thought. Do you think it would be possible to progamatically generate the actual message code if we had the BNF for the Kafka protocol? I think this was originally suggested by @edenhill. Supposedly, the Java client uses https://github.com/klarna/kafka_protocol somehow. I'm wondering if we could actually generate all of the per-message stubs using a python script? Thoughts?

adamkotwasinski commented:
@mattklein123 Yes, this is very much possible.
What's done in https://github.com/klarna/kafka_protocol  is basically using internal class of kafka client and creating BNF-like https://github.com/klarna/kafka_protocol/blob/master/priv/kafka.bnf 
Then this spec gets translated into Erlang : https://github.com/klarna/kafka_protocol/blob/master/src/kpro_schema.erl#L159

given the bnf file, it should be relatively easy to create corresponding c++ code (there might be some drawbacks that we might end up with concrete classes like `FetchRequestV1`, `FetchRequestV2` as they might differ in fields - but that should be acceptable (Erlang doesn't appear to have that problem as that's all tuples like Prolog)

_fun factor_ : the generated code might _really_ end up looking similar to this: https://github.com/envoyproxy/envoy/pull/4950#discussion_r234663642

mattklein123 commented:
@adamkotwasinski great, huge +1 if we can get the stubs auto generating. If so, TBH, I care much less about all the template foo or other ugliness that is required to make it work. Up to you whether you want to take a crack at this in this PR, or just agree on what a single message looks like end to end and then work on generation of the rest of the messages? I think this is going to be well worth it in overall readability and maintenance.

adamkotwasinski commented on source/extensions/filters/network/kafka/kafka_protocol.h:
```diff
@@ -0,0 +1,25 @@
+ * Kafka request type identifier (int16_t value present in header of every request)
+ * @see http://kafka.apache.org/protocol.html#protocol_api_keys
+ */
+enum RequestType : int16_t {
```
Can I keep it this way?
Making enum class out of request type will require adding another value for "unknown request" and possibly forces us to keep two fields in `UnknownRequest` - `RequestType` for `unknown value` and then `int16_t` with underlying 

The other idea would be to make N constexprs like
```
constexpr int16_t PRODUCE_REQUEST_TYPE{0}
....
constexpr int16_t OFFSET_COMMIT_REQUEST_TYPE{8}
```

extemely related : https://stackoverflow.com/questions/1965249/how-to-write-a-java-enum-like-class-with-multiple-data-fields-in-c

edenhill commented:
Would it be useful if librdkafka provided an API to send/receive arbitrary request/responses blobs to logical (e.g., partition leaders, group coordinators, etc) or actual (by address or broker id) brokers?
This would allow Envoy to do all its intercepting glory without having to worry about the logical aspects of Kafka, nor the security protocols.

Implementing a full Kafka client and keeping it up to date is a big commitment, and this approach would lessen that burden.

adamkotwasinski commented:
@edenhill A lot.
I think what'd be missing are support for other request/response types (like `StopReplica` that flows only between brokers AFAICT). And the serialization of responses / deserialization of requests, as you have mentioned.

mattklein123 commented:
@edenhill I think this came up when we first chatted 6 months ago or so. In an optimal world, we would be able to use librdkafka to do all of the protocol parsing and serialization, and then allow Envoy to do the actual networking in terms of where to send packets, etc. I'm not familiar with librdkafka to know how that would work, but I thought you said that something like this wouldn't be possible. Is that what you are suggesting?

adamkotwasinski commented:
I'd also one thing here:
"rich" librdkafka's features are needed to have a real kafka-cluster proxy, but not for trivial filter that basically collects metrics (which is my initial intent, but still needs the parse, hence this PR)

edenhill commented:
@mattklein123 Right, what we discussed was if librdkafka could provide a high level API to the protocol requests, which it wasn't really designed for.  What I'm suggesting here is somewhat different, namely that librdkafka simply provides a socket abstraction after the initial Kafka&security handshake has taken place. Envoy would still (de)serialize the requests going between the client and broker, but would not have to bother with the stateful parts of the Kafka protocol: leaders, coordinators, security, etc.
Envoy would focus on the what rather than the how. So somewhat of a middle ground.

What this would look like in detail depends on what Envoy is planning to do with the intercepted Kafka traffic. Is there a desired-features list somewhere I can read to know if this would work in practice?

mattklein123 commented:
@edenhill there is no design doc yet (there should be). For a laundry list of features people are asking for, see here: https://github.com/envoyproxy/envoy/issues/2852

At a high level, I see 2 major use cases:
1) Bump in the wire interception for stats, tracing, logs, etc. This would mostly be useful broker side stats gathering, though it could probably be used if there was protocol inference and we transparently redirected all client traffic through Envoy such that all broker connections end up going through Envoy.

2) Allow the client Kafka client to be connected to Envoy as the broker. Envoy then becomes the real kafka client, and then allows load balancing, health checking, etc., and the in-process client just knows about a single Envoy "logical broker." Obviously this case is really complicated but something that I would like to make happen.

Optimally, we would not reimplement all of the stuff in librdkafka since IMO much of it would be useful. Is it worth another meeting to chat about this? I can work on scheduling offline.

mattklein123 commented:
@adamkotwasinski FWIW I think we should still push forward with this PR as we have discussed (with message auto generation from the BNF). In parallel I will kick off a thread with @edenhill and we can discuss other options. Sound good?

adamkotwasinski commented:
@mattklein123 Agree. Will come back with some commits after new year (unless `librdkafka` adoption will prove to be more viable :) )

mattklein123 commented:
@adamkotwasinski sorry for the delay, I will look through this soon.

adamkotwasinski commented on source/extensions/filters/network/kafka/generated/kafka_request_resolver.cc:
```diff
@@ -0,0 +1,36 @@
+// DO NOT EDIT - THIS FILE WAS GENERATED
```
@mattklein123 Actual decision is yours, as then you are requiring to have jvm & gradle installed (as you need to get Kafka's Java library to get the protocol data, so you need Java - there is *no* BNF with Kafka spec, the BNF was created from Java code).

Erlang https://github.com/klarna/kafka_protocol#code-generation chose the other way - the generated files are committed.

It's up to you if you want to pull java&related tooling into your build ecosystem - I'm fine either way.

mattklein123 replied:
We use Bazel. It's there. Yes I would prefer to make it hermetic. I think this can likely be done as an external genrule Bazel target.

gwenshap replied:
You may be interested in https://github.com/apache/kafka/pull/5893
Where we added a real protocol-generator to Apache Kafka. It will take us a while to get all our request/response pairs, but it will be great if we'll generate from same definitions.

adamkotwasinski replied:
so with request/resp spec in json, we can remove java from ecosystem
I'll make rewrite the existing generator to python and invoke it as part of build

adamkotwasinski replied:
@gwenshap FYI - if this gets merged in, we will depend on the fact that spec files present at https://github.com/apache/kafka/tree/2.2/clients/src/main/resources/common/message are named `*Request.json` (and in future `*Response.json` too)

adamkotwasinski commented:
/wait

adamkotwasinski commented:
/wait

adamkotwasinski commented:
/wait

adamkotwasinski commented:
/wait

adamkotwasinski commented:
```
ERROR: Error while writing profile file: No space left on device
```

adamkotwasinski commented on api/bazel/repositories.bzl:
```diff
@@ -16,6 +16,10 @@ GOOGLEAPIS_SHA = "16f5b2e8bf1e747a32f9a62e211f8f33c94645492e9bbd72458061d9a9de1f
 PROMETHEUS_GIT_SHA = "99fa1f4be8e564e8a6b613da7fa6f46c9edafc6c"  # Nov 17, 2017
 PROMETHEUS_SHA = "783bdaf8ee0464b35ec0c8704871e1e72afa0005c3f3587f65d9d6694bf3911b"
 
+OPENCENSUS_GIT_SHA = "ab82e5fdec8267dc2a726544b10af97675970847"  # May 23, 2018
```
after merge

adamkotwasinski commented on api/bazel/repositories.bzl:
```diff
@@ -16,6 +16,10 @@ GOOGLEAPIS_SHA = "16f5b2e8bf1e747a32f9a62e211f8f33c94645492e9bbd72458061d9a9de1f
+OPENCENSUS_GIT_SHA = "ab82e5fdec8267dc2a726544b10af97675970847"  # May 23, 2018
+OPENCENSUS_SHA = "1950f844d9f338ba731897a9bb526f9074c0487b3f274ce2ec3b4feaf0bef7e2"
+
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_file")
```
will be removed when kafka open source releases 2.3 client jar

adamkotwasinski commented on api/bazel/repositories.bzl:
```diff
@@ -31,6 +35,16 @@ def api_dependencies():
         strip_prefix = "protoc-gen-validate-" + PGV_RELEASE,
         url = "https://github.com/lyft/protoc-gen-validate/archive/v" + PGV_RELEASE + ".tar.gz",
     )
+    http_file(
```
will be replaced with http_archive or something similar to download kafka 2.3 client jar, then it will be unpacked to get all kafka request `.json` spec files

adamkotwasinski commented on source/extensions/filters/network/kafka/protocol_code_generator/kafka_request_resolver_cc.j2:
```diff
@@ -0,0 +1,28 @@
+ParserSharedPtr RequestParserResolver::createParser(int16_t api_key, int16_t api_version,
+                                                    RequestContextSharedPtr context) const {
+
+{% for request_type in request_types %}{% for field_list in request_type.compute_field_lists() %}
```
resolves apikey * apiversion -> parser

mattklein123 replied:
Please add comments. :)

adamkotwasinski commented on source/extensions/filters/network/kafka/protocol_code_generator/requests_test_cc.j2:
```diff
@@ -0,0 +1,61 @@
+// DO NOT EDIT - THIS FILE WAS GENERATED
```
automated tests so serialization changes are actually tested on live data

mattklein123 replied:
Please add as a comment, also similar comments about where this file ends up being generated to (same elsewhere)

adamkotwasinski commented:
@mattklein123 Kafka request classes now generated from Kafka jsons originating at Kafka source repo (this will be changed when 2.2 finally releases as jar)

adamkotwasinski commented:
/nowait

adamkotwasinski commented on source/extensions/filters/network/kafka/serialization.h:
```diff
@@ -0,0 +1,776 @@
+ * For non-null values, first the length N is given as an INT32. Then N bytes follow.
+ * A null value is encoded with length of -1 and there are no following bytes.
+ */
+class NullableBytesDeserializer : public Deserializer<NullableBytes> {
```
honestly, `NullableBytesDeserializer` & `BytesDeserializer` (or `N~StringD~`, or `N~ArrayD~`) are so similar that they could inherit from a single class with some minor configuration object (strategy to tell us what to do with given `int length`) and output converter (if we store result as `optional<vector<T>` then BytesDeserializer needs to invoke unpack the option, while the NullableB~D~ does not
I'll wait if I've caused any worse errors first, then happy to change it if requested

mattklein123 replied:
Up to you. In general this all looks great, so however you want to wind up, though I would personally favor less duplicated code if possible.

mattklein123 commented on source/extensions/filters/network/kafka/kafka_request.h:
```diff
@@ -0,0 +1,99 @@
+  UnknownRequest(const RequestHeader& request_header) : AbstractRequest{request_header} {};
+
+  // this isn't the prettiest, as we have thrown away the data
+  // XXX discuss capturing the data as-is, and simply putting it back
```
s/XXX/TODO per style guide.

mattklein123 commented on source/extensions/filters/network/kafka/kafka_request_parser.h:
```diff
@@ -0,0 +1,197 @@
+  /**
+   * Request parser singleton
+   */
+  static const RequestParserResolver INSTANCE;
```
nit: s/INSTANCE/instance, but better, make this a function and use CONSTRUCT_ON_FIRST_USE.

mattklein123 commented on source/extensions/filters/network/kafka/kafka_request_parser.h:
```diff
@@ -0,0 +1,197 @@
+    try {
+      context_->remaining_request_size_ -= deserializer.feed(buffer, remaining);
+    } catch (const EnvoyException&) {
+      // treat the whole request as invalid, throw away the rest of the data
```
Question: Does this happen normally? If not, should we avoid trying to be smart and skipping a request vs. just letting this type of exception propagate fully out and we can kill the connection or just stop trying to parse like we do with Mongo?

adamkotwasinski replied:
No, it does not (it means that client is sending garbage or deserialization is buggy).
Will propagate the exception - the filter will know best what to do with it (metrics filter can just disable sniffing; thick kafka-facade should stop iteration I think).

mattklein123 commented on source/extensions/filters/network/kafka/parser.h:
```diff
@@ -0,0 +1,82 @@
+   * @param remaining remaining data in buffer, will be updated by parser
+   * @return parse status - decision what should be done with current parser (keep/replace)
+   */
+  virtual ParseResponse parse(const char*& buffer, uint64_t& remaining) PURE;
```
Ping on this comment?

mattklein123 commented on source/extensions/filters/network/kafka/protocol_code_generator/complex_type_template.j2:
```diff
@@ -0,0 +1,37 @@
+struct {{ complex_type.name }} {
```
Can we have some comments at the top of each template file explaining roughly what the template is for? It can be hard to dive into the template code otherwise.

mattklein123 commented on source/extensions/filters/network/kafka/protocol_code_generator/kafka_generator.py:
```diff
@@ -0,0 +1,422 @@
+#!/usr/bin/python
+
+# usage:
```
If possible can you add some high level comments of what this does and why?
1) Where are definitions pulled from
2) What templates we use
3) What the output is, etc.

mattklein123 commented on source/extensions/filters/network/kafka/protocol_code_generator/kafka_request_resolver_cc.j2:
```diff
@@ -0,0 +1,28 @@
+// DO NOT EDIT - THIS FILE WAS GENERATED
```
Where are you generating this code? I don't think this should be necessary? Can you take a look at where we generate protobuf files to? I think those go into Bazel internal directories so we shouldn't need to worry about these type of comments or clang-format? Also, same comment about comments on the template file.

mattklein123 commented on source/extensions/filters/network/kafka/protocol_code_generator/kafka_request_resolver_cc.j2:
```diff
@@ -0,0 +1,28 @@
+namespace NetworkFilters {
+namespace Kafka {
+
+const RequestParserResolver RequestParserResolver::INSTANCE;
```
nit: similar CONSTRUCT_ON_FIRST_USE comment.

mattklein123 commented on source/extensions/filters/network/kafka/serialization.h:
```diff
@@ -0,0 +1,776 @@
+ * buffer/remaining)
+ * @param T type of deserialized data
+ */
+template <typename T> class Deserializer {
```
I still do wonder if this code has overlaps with what we are doing with Thrift/MySQL/etc. I feel like we end up rewriting variants of this each time. I don't know if that's exactly true here or not, but it would be nice to one day figure that out. This is probably non-actionable but it might be worth a quick look and a TODO. cc @zuercher @rshriram

mattklein123 commented on source/extensions/filters/network/kafka/serialization.h:
```diff
@@ -0,0 +1,776 @@
+ * structure-tree during encoding (currently api_version, as different request versions serialize
+ * differently)
+ */
+// XXX that class might be split into Request/ResponseEncodingContext in future,
```
nit: s/XXX/TODO same elsewhere if I didn't catch anyway.

mattklein123 commented on source/extensions/filters/network/kafka/serialization_composite.h:
```diff
@@ -0,0 +1,468 @@
+// XXX this file can be generated, as it's repeating the same code for 0..9 delegates
```
You already have all of the jinja setup. Do you want to just do this now? It would also make it simpler to review.

adamkotwasinski commented:
/wait

adamkotwasinski commented:
/wait

adamkotwasinski commented:
/wait

adamkotwasinski commented:
Kafka spec is downloaded from `2.2.0-rc2` release at https://github.com/apache/kafka/releases - the interesting spec files are then fed into generator scripts.
More tests (including unit & integration ones for request_codec.h have been provided (some are generated)).
Replaced char pointers with absl string_view.
Composite serializer header and related tests are now generated (via separate script).
Added documentation.

Remaining (non-critical) work imho is:
1) add `size()` method to requests (so we do not need a temporary buffer to compute serialized request's size)
2) unifying code in (Nullable)Bytes/String/Array deserializers
3) replace some shared pointers with unique pointers in places like passing pointer (with response/parser) as a result

@mattklein123 FYI

mattklein123 commented:
@adamkotwasinski from this point forward can we avoid force pushing? I'm going to review now. I think we are close and if there are further changes it will allow me to review easily. Thank you.

mattklein123 commented on source/extensions/filters/network/kafka/request_codec.cc:
```diff
@@ -0,0 +1,90 @@
+namespace Kafka {
+
+class RequestStartParserFactory : public InitialParserFactory {
+
```
nit: del newline, similar elsewhere.

adamkotwasinski commented on source/extensions/filters/network/kafka/message.h:
```diff
@@ -0,0 +1,33 @@
+   * Encode the contents of this message into a given buffer
+   * @param dst buffer instance to keep serialized message
+   */
+  virtual size_t encode(Buffer::Instance& dst) const PURE;
```
No, it cannot.
The buffer is modified by calling `dst.add` several times, when I actually put the message's contents inside.

adamkotwasinski commented on source/extensions/filters/network/kafka/parser.h:
```diff
@@ -0,0 +1,83 @@
+   * @param data bytes to be processed, will be updated by parser if any have been consumed
+   * @return parse status - decision what should be done with current parser (keep/replace)
+   */
+  virtual ParseResponse parse(absl::string_view& data) PURE;
```
right now `data` is modified, in a fashion similar to `const char*& data, size_t& remaining` before

adamkotwasinski commented:
/wait

adamkotwasinski commented:
/wait got some problems running coverage reports locally, might end up creating some noise in here

adamkotwasinski commented:
/wait

adamkotwasinski commented:
/wait not yet

adamkotwasinski commented:
/wait

adamkotwasinski commented:
/wait

adamkotwasinski commented:
/wait

adamkotwasinski commented:
/retest

mattklein123 commented on source/extensions/filters/network/kafka/kafka_request.h:
```diff
@@ -82,15 +79,16 @@ template <typename RequestData> class ConcreteRequest : public AbstractRequest {
-  // TODO(adamkotwasinski) discuss capturing the data as-is, and simply putting it back
-  // this would add ability to forward unknown types of requests in cluster-proxy
+  /**
+   * It is impossible to encode unknown request, as it is only a placeholder.
```
So this can never happen in practice? If so should it be NOT_IMPLEMENTED?

adamkotwasinski commented:
/wait
I'm afraid it might be something happening due to gcovr not handling the python-generated files properly, will investigate that, but it might take some time.

adamkotwasinski commented:
/retest

adamkotwasinski commented:
the coverage should be fixed now but I still need to address the 100% coverage & other comments
/wait

adamkotwasinski commented:
/wait

adamkotwasinski commented:
/retest

adamkotwasinski commented:
@mattklein123 
Coverage fixed (generated files are put in `external` directory what makes them ignored by gcovr).
Made parser/codec classes templated to prepare for future response support.
Introduced new type `RequestParseFailure` to carry information about failed parses (most often that would be newer client sending message that wouldn't be known to envoy).
Got rid of `message.h` as right now I see no point in having a high-level super class for both request & response - it's like the only thing they have in common today is `encode` method.
